### PR TITLE
Http transport -- DO NOT MERGE

### PR DESCRIPTION
**Pull request by hasLeland**
DO NOT MERGE

This pull request exists purely for code review reasons. Rip this code apart!

Summary:

Client and server can now communicate with each other over HTTP. Parameters can be specified within the query and the path, and they'll be transmitted in the query and path. The body still works too.

Also, I re-wrote how the internals of several types within doctree work so that they're easier to work with in external packages. However, nothing else about them changed.

### CODE REVIEW

**Comment by hasLeland** [comment]
I made this pull request because the prior one was for the wrong branch, and I've now rebased things together.

File: protoc-gen-truss-gokit/generator/httptransport/embeddable_funcs.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+	parts := strings.Split(url, "/")
+	for k, v := range pmp {
+		rv[k] = parts[v]
+	}
+
+	return rv, nil
+}
+
+// Given a url template, create a map of the
```
Comment (hasAdamr) [inline protoc-gen-truss-gokit/generator/httptransport/embeddable_funcs.go:33]: of the what?

File: protoc-gen-truss-gokit/generator/httptransport/get_source.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+	"go/ast"
+	"go/parser"
+	"go/printer"
+	"go/token"
+)
+
+// FuncSourceCode returns a string representing the source code of the function
+// provided to it.
+func FuncSourceCode(val interface{}) (string, error) {
```
Comment (hasAdamr) [inline protoc-gen-truss-gokit/generator/httptransport/get_source.go:18]: Might want to add a little more to this comment stating that it requires the source of truss to be on disk to work.

File: protoc-gen-truss-gokit/generator/httptransport/httptransport.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+type Helper struct {
+	Methods           []*Method
+	PathParamsBuilder string
+}
+
+// NewHelper builds a helper struct from a service declaration. The other
+// "New*" functions in this file are there to make this function smaller and
+// more testable.
+func NewHelper(svc *doctree.ProtoService) *Helper {
+	pp, _ := FormatCode(HTTPAssistFuncs)
```
Comment (hasAdamr) [inline protoc-gen-truss-gokit/generator/httptransport/httptransport.go]: A comment about this global value existing in `httptransport/templates.go` would be good. Possibly define `HTTPAssistFuncs` immediately before this function and comment that it's components are defined in `templates.go`

File: protoc-gen-truss-gokit/generator/httptransport/httptransport_test.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+	for _, test := range cases {
+		got := EnglishNumber(test.i)
+		if got != test.want {
+			t.Errorf("Got %v, want %v\n", got, test.want)
+		}
+	}
+}
+
+func TestLowCamelName(t *testing.T) {
```
Comment (hasAdamr) [inline protoc-gen-truss-gokit/generator/httptransport/httptransport_test.go:203]: I really like this test

File: gendoc/doctree/docs.go [hunk 5]
```diff
@@ -185,14 +175,33 @@ func (self *MicroserviceDefinition) Markdown() string {
...
-	Services []*ProtoService
+	Describable
+	Name        string
+	Description string
+	Messages    []*ProtoMessage
+	Enums       []*ProtoEnum
+	Services    []*ProtoService
+}
+
+func (self *ProtoFile) GetName() string {
```
Comment (hasIan) [inline gendoc/doctree/docs.go:186]: See the [section of _Effective Go_ on **Getters**](https://golang.org/doc/effective_go.html#Getters) for naming conventions (spoiler: no `Get` prefix).

File: gendoc/doctree/docs.go [hunk 6]
```diff
@@ -185,14 +175,33 @@ func (self *MicroserviceDefinition) Markdown() string {
...
+	return self.Name
+}
+
+func (self *ProtoFile) SetName(s string) {
```
Comment (hasIan) [inline gendoc/doctree/docs.go:190]: [Don't use `self` or `this` as receiver names.](https://github.com/golang/go/wiki/CodeReviewComments#receiver-names) Golint actually complains about this. Conventionally, an abbreviated name of the type is used. E.g., `pf` here.

**Comment by hasLeland** [comment]
After chatting in person, there are some inconsolable issues with the design of Doctree and the associated `Describable` interface that leads to the stated oddness with getters and setters being named the way they are, and my plan is to not address them in this pull request but to wait till tomorrow or the next day to begin a doctree refactor.

**Pull request closed without merging**